Build fails with "Can't resolve ...@embroider/macros..."

Example: https://github.com/ember-learn/guidemaker-ember-template/actions/runs/5416696506/jobs/9857850370#step:5:82

https://discord.com/channels/480462759797063690/568935504288940056/1124312264204374177

```
ERROR in ../rewritten-packages/ember-svg-jar.e02db158/_app_/helpers/svg-jar.js 1:0-102
Module not found: Error: Can't resolve '../../../../../@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat' in '/home/runner/work/guidemaker-ember-template/guidemaker-ember-template/node_modules/.embroider/rewritten-app/helpers'
resolve '../../../../../@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat' in '/home/runner/work/guidemaker-ember-template/guidemaker-ember-template/node_modules/.embroider/rewritten-app/helpers'
  using description file: /home/runner/work/guidemaker-ember-template/guidemaker-ember-template/node_modules/.embroider/rewritten-app/package.json (relative path: ./helpers)
    Field 'browser' doesn't contain a valid alias configuration
    No description file found in /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon or above
    no extension
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat doesn't exist
    .wasm
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.wasm doesn't exist
    .mjs
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.mjs doesn't exist
    .js
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.js doesn't exist
    .json
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.json doesn't exist
    .ts
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.ts doesn't exist
    .hbs
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.hbs doesn't exist
    .hbs.js
      Field 'browser' doesn't contain a valid alias configuration
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat.hbs.js doesn't exist
    as directory
      /home/runner/work/guidemaker-ember-template/@embroider/compat/node_modules/@embroider/macros/src/addon/es-compat doesn't exist
 @ ./assets/dummy.js 535:13-48
```

--- a/packages/core/src/module-resolver.ts
+++ b/packages/core/src/module-resolver.ts
@@ -802,6 +802,15 @@ export class Resolver {
       // a compat adapter). In the metadata, they would be listed in
       // package-relative form, so we need to convert this specifier to that.
       let absoluteSpecifier = resolve(dirname(fromFile), specifier);
+
+      if (!absoluteSpecifier.startsWith(pkg.root)) {
+        // this relative path escape its package. So it's not really using
+        // normal inter-package resolving and we should leave it alone. This
+        // case comes up especially when babel transforms are trying to insert
+        // references to runtime utilities, like we do in @embroider/macros.
+        return logTransition('beforeResolve: relative path escapes its package', request);
+      }
+
       let packageRelativeSpecifier = explicitRelative(pkg.root, absoluteSpecifier);
       if (isExplicitlyExternal(packageRelativeSpecifier, pkg)) {
         let publicSpecifier = absoluteSpecifier.replace(pkg.root, pkg.name);

--- a/tests/scenarios/core-resolver-test.ts
+++ b/tests/scenarios/core-resolver-test.ts
@@ -479,6 +479,25 @@ Scenarios.fromProject(() => new Project())
             .to('./node_modules/my-addon/_app_/hello-world.js');
         });
 
+        test('app-js module in addon can still do relative imports that escape its package', async function () {
+          givenFiles({
+            'node_modules/extra.js': '',
+            'node_modules/my-addon/_app_/hello-world.js': 'import "../../extra.js"',
+            'app.js': `import "my-app/hello-world"`,
+          });
+
+          await configure({
+            addonMeta: {
+              'app-js': { './hello-world.js': './_app_/hello-world.js' },
+            },
+          });
+
+          expectAudit
+            .module('./node_modules/my-addon/_app_/hello-world.js')
+            .resolves('../../extra.js')
+            .to('./node_modules/extra.js');
+        });
+
         test('hbs in addon is found', async function () {
           givenFiles({
             'node_modules/my-addon/_app_/templates/hello-world.hbs': '',

